### PR TITLE
Make job name link to testgrid.

### PR DESCRIPTION
**Pull request by rmmh**
Plus some other misc fixes.

image
image

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/658)

### CODE REVIEW

File: gubernator/templates/build_list.html [hunk 1]
```diff
@@ -4,21 +4,27 @@
 <h1><a href="/jobs{{job_dir|dirname|dirname}}">{{job_dir|dirname|dirname}}</a>/{{job}}</h1>
 % endblock
 % block content
+<div id="content">
+% if testgrid_query
+<a href="{{testgrid_query | tg_url}}"><h3>Testgrid History for Job</h3></a>
```
Comment (fejta) [inline gubernator/templates/build_list.html]: Awesome! Maybe have it link to a github issue to request we add support for this job if it isn't there?? :)

File: gubernator/view_build.py [hunk 2]
```diff
@@ -129,16 +129,57 @@ def get(self, prefix, job, build):
...
+    try:
+        if 'pr-logs' in job_dir:
+            raise ValueError('bad code path for PR build list')
+        # If we have latest-build.txt, we can skip an expensive GCS ls call!
+        latest_build = int(latest_fut.get_result())
+        if before:
+            latest_build = int(before) - 1
+        builds = range(latest_build, max(0, latest_build - 40), -1)
+    except (ValueError, TypeError):
+        fstats = view_base.gcs_ls(job_dir)
```
Comment (fejta) [inline gubernator/view_build.py]: nit: sorted(view_base...)?

File: gubernator/view_build.py [hunk 3]
```diff
@@ -129,16 +129,57 @@ def get(self, prefix, job, build):
...
+    Given a job dir, give a (partial) list of recent build
+    finished.jsons.
+
+    Args:
+        job_dir: the GCS path holding the jobs
+    Returns:
+        a list of [(build, finished)]. build is a string like "123",
+        finished is either None or a dict of the finished.json.
+    '''
+    latest_fut = gcs_async.read(job_dir + 'latest-build.txt')
```
Comment (fejta) [inline gubernator/view_build.py]: What is a fut?

File: gubernator/view_build.py [hunk 4]
```diff
@@ -129,16 +129,57 @@ def get(self, prefix, job, build):
...
+        builds = [os.path.basename(os.path.dirname(f.filename))
+                  for f in fstats if f.is_dir]
+        if before and before in builds:
+            builds = builds[builds.index(before) + 1:]
+        builds = builds[:40]
+    finished_futs = {build: gcs_async.read(
+        '%s%s/finished.json' % (job_dir, build))
+        for build in builds}
+
+    def from_json(fut):
```
Comment (fejta) [inline gubernator/view_build.py]: nit: consider a more descriptive name

File: gubernator/view_build.py [hunk 5]
```diff
@@ -129,16 +129,57 @@ def get(self, prefix, job, build):
             failures=failures, build_log=build_log, pr=pr, pr_digest=pr_digest,
             testgrid_query=testgrid_query))
 
+@view_base.memcache_memoize('build-list://', expires=60)
+def build_list((job_dir, before)):
```
Comment (fejta) [inline gubernator/view_build.py]: What does the double paren do? build_list accepts 1 arg: a 2 part tuple??

File: gubernator/templates/build.html [hunk 6]
```diff
@@ -7,37 +7,41 @@
...
+	% else
+		{{job}}
 	% endif
+	#{{build}}</h1>
+	<p><a href="/builds{{job_dir}}">Recent runs</a><br>
 % endblock
 % block content
 <div id="summary">
+	<table border=0 align="center" style="text-align: left;">
+	<style>tr:nth-child(even){background:#ddd}</style>
```
Comment (fejta) [inline gubernator/templates/build.html]: Does that mean this style only applies to inside the table tag??

File: gubernator/view_build.py (hunk 3 above)
Comment (rmmh) [inline gubernator/view_build.py]: A future. It's used in a few other places in this file too.

File: gubernator/view_build.py (hunk 5 above)
Comment (rmmh) [inline gubernator/view_build.py]: It's a weird side-effect of how the memcache_memoize decorator works-- it doesn't support more than one argument. This should be cleaned up in a future refactor.

File: gubernator/templates/build_list.html (hunk 1 above)
Comment (rmmh) [inline gubernator/templates/build_list.html]: That should be tested automatically Real Soon Now, so I'm not going to worry about putting the functionality here.

File: gubernator/view_build.py (hunk 4 above)
Comment (rmmh) [inline gubernator/view_build.py]: Done.

File: gubernator/templates/build.html (hunk 6 above)
Comment (rmmh) [inline gubernator/templates/build.html]: Yes, style tags are scoped to the enclosing element.

**Pull request merged**